### PR TITLE
Navier stokes solvers for multiple devices

### PR DESCRIPTION
**Pull request by mikaem**
## Summary

The `KMM2D` channel solver and its `RayleighBenard` subclass were accurate and fast on a single
device but crashed under `mpirun`, and nothing in the library was actually distributed: the
compiled step contained **8 all-gathers per step**, so every device held the whole field and did
the whole solve.

This branch makes both demos run correctly on one device, on several devices in one process, and
under `mpirun` with 1, 2 or 4 processes — with no all-gathers, and *faster* on a single device
than before.

The last blocker was that a real half-spectrum stores `N/2 + 1` coefficients, which is odd for
every power-of-two `N`, so difficult for device count to split it. Choosing `N` from the
coefficient count instead does not scale: `rfft(254)` costs 160 µs against 31 µs for `rfft(256)`,
because 127 is prime. `RFourier` now pads its stored axis instead, so FFT-friendly sizes and
shardable sizes stop being disjoint.

## Types of Changes

- [x] Bug fix
- [x] New feature
- [x] Performance improvement
- [ ] Documentation update
- [x] Refactor / cleanup
- [ ] Build / CI
- [ ] Other (describe):

## Related Issues

Fixes # \
Refs #58

## Description of Changes

**`RFourier` pads its coefficient axis** (`src/jaxfun/galerkin/Fourier.py`). `RFourier(N)` keeps
`n_real = N/2 + 1` wavenumbers and stores `n_real + n_extra`, where `n_extra` defaults to whatever
the device count needs — zero on one device, so single-device shapes are untouched. The padding
slots are structurally zero: `forward`/`scalar_product` zero-fill them, `backward` slices them off
before `irfft`, `hermitian_weights` is 0 there so they are the zero function, and their wavenumber
is the Nyquist rather than 0 so their operator block stays invertible (at 0 a pure Laplacian on a
fully periodic space would return NaN there). `num_dofs` is the padded count, which is what keeps
the change small — all 56 `num_dofs` call sites work unchanged.

**`Chebyshev` transforms** now make use of new and locally optimized dct/idct functions that are considerably  faster than `jax.scipy.fft` versions. 

**The transforms actually distribute.** The eight all-gathers were not in the transforms at all:
`TensorProductSpace.to_orthogonal` applies each axis's stencil with `rmatvec`, and on a Fourier
axis that stencil is already a `DiagonalMatrix` — but `DiagonalMatrix` overrode `matvec` and not
`rmatvec`, so the call fell through to the banded implementation, which pads by `m - 1` on each
side of the contracted axis. On the split axis that padding straddles the split and the field is
gathered onto every device. One method (`DiagonalMatrix.rmatvec`, delegating to `matvec`) removed
all eight.

**One jit boundary for the integrators** (`src/jaxfun/integrators/base.py`). `_advance` is now the
only one; `step` is a batch of one and `solve` a batch of many, so the two cannot compile
different things. This exposed three ordinary bugs: `setup()` rebuilding operators behind an
identity-keyed cache, `solve` passing static arguments by keyword (where `static_argnums` has no
effect), and `warm_operator_solve_cache` swallowing real failures.

**The wavenumber solver optimized for multi-device** (`src/jaxfun/la/tpmatrix.py`). Both GPU and CPU. For both the odd/even parity property of the 1D matrices is taken advantage of in batch. For GPU the default LU substitution method is chosen as a cyclic associative_scan (`prefix`) instead of scan. This is faster on gpu, but slower on cpu. For cpu regular scan is used by default. New environment variable `JAXFUN_LU_SUBSTITUTION_ALGORITHM` = `scan`, `prefix` or `auto` makes it possible to choose.

**Batched transforms now shard.** `backward_batch` and friends used to refuse a distributed input.
`_apply_separable_spmd_shard_map` takes a `batch_dims` argument so the batch axis is excluded from
the `all_to_all` role assignment; one `all_to_all` serves the whole batch rather than one per
field.

**`examples/spmd_bootstrap.py`** (new) brings up `jax.distributed`, and carries `is_leader`,
`echo` (leader-only `print`) and `to_host` (gather for plotting). It imports nothing from `jaxfun`
at module scope, because `jaxfun.sharding` builds its device mesh at import time.

Two ways in. `initialize_distributed` is the generic one and is what the demos call: it detects an
MPI launcher from the environment (`OMPI_COMM_WORLD_SIZE`, `PMI_SIZE`, `PMIX_RANK`), then takes
the world size and rank from `MPI.COMM_WORLD` itself, since every launcher spells those
differently. `initialize_saga` is the Slurm counterpart, for clusters where the job script defines
the world rather than `mpirun`; its docstring states which variables such a script has to export
(`JAX_PLATFORM`, `LOCAL_DEVICES`, alongside Slurm's own). It is deliberately *not* wired into
`initialize_distributed`: Slurm sites differ enough that a user is going to write their own launch
script regardless, and a concrete example they can copy and edit beats a detection heuristic that
guesses wrong on the next cluster.

## Screenshots / Logs (if applicable)

`examples/RayleighBenard.py` at its defaults (`Ra = 1e6`, `Pr = 0.7`, `M, N = 128, 64`), the
final frame at `t = 60`. Identical on 1, 2 and 4 processes — the animation comes out
byte-for-byte the same.

Rayleigh-Benard convection at Ra=1e6, Pr=0.7: temperature field with the velocity quiver below

Collectives in the compiled `KMM2D` step, and single-device throughput:

```
before:  8 all-gather, 4 all-reduce                      554-576 steps/s (M=34)
after:   0 all-gather, 8 all-to-all, 4 all-reduce        607-610 steps/s (M=34)
                                                          675.6 steps/s (M=32, now available)
```

Agreement across configurations:

```
OrrSommerfeld (M=32)     serial 5.07e-07 | mpirun -np 2/-np 4 5.06e-07 | 4 in-process 5.06e-07
RayleighBenard (M=128)   dofs (65,.) / (66,.) / (68,.) on 1 / 2 / 4 processes
                         every diagnostic identical; byte-identical rayleighbenard.gif
```

## Breaking Changes

- [x] Yes (describe below)
- [ ] No

Two, both only visible on a multi-device host:

1. **`RFourier(N).N` depends on the device count.** It is `N/2 + 1` on one device (unchanged) and
   padded up to a multiple of the device count on more. A coefficient array saved from a
   1-device run therefore has fewer rows than a 2-device run expects. Converting is a slice or a
   zero-pad, since the tail is always zero. Pass `n_extra=0` to restore the old shapes, at the
   cost of not being able to distribute the axis.
2. **`TensorProductSpace._use_spmd` decides from shapes, not from array placement.** It has to:
   `Tracer.devices()` raises, and these transforms must be callable from inside a jitted step. On
   a multi-device host a space whose extents divide is therefore distributed even if the caller
   handed it a local array, and the result comes back carrying a sharding. Sizes that do not
   divide run locally as before.

## Checklist

- [x] Tests added or updated
- [x] Documentation updated (README, examples, docstrings)
- [x] Added type hints
- [x] Linting & formatting pass locally (`pre-commit run --all-files`)
- [x] All new and existing tests pass (`uv run pytest`)
- [ ] Coverage does not decrease
- [x] Git history is clean (squash/fixup before merge)

`pytest tests` 1137 passed / 8 skipped; `-m spmd --num-devices=2` and `=4` 58 passed each;
`-m examples` 30 passed / 3 skipped; `ruff` and `ty` clean.

## Additional Notes

Review of this branch turned up four issues, all fixed here:

- **`_use_spmd` raised where it should have degraded.** Scoping the raise to "the axis split
  across devices" was still too broad: in the physical direction that axis is the wall-normal
  *quadrature* count, which the space fixes for itself and no caller can pad. `RayleighBenard(32,
  16, ...)` on 3 devices therefore died in its constructor, on a one-off scalar reduction used
  only to compute `self.vol`. The raise is now scoped to the one case that is actionable — a
  half-spectrum space whose leading axis does not divide, which is exactly what `RFourier`'s
  `n_extra` exists to pad. Everything else runs locally, losing parallelism and nothing else.
  Verified: the solver now constructs and steps to identical results on 2, 3, 4 and 5 devices.
- **`_check_shardable` tested the wrong quantity** — `n_F`, the product of all Fourier extents,
  where `spectral_sharding` splits `shape[0]` alone. For a 3-D operator the two differ, so the
  check passed and the failure resurfaced later. It now tests axis 0, as its own message always
  claimed.
- **`inner.py` forced a sharding without checking it divides**, at both placement sites. They now
  go through `place`, which leaves an array the mesh cannot divide alone — `IndivisibleError` is a
  `ValueError`, so that path already handled it and the two sites were simply bypassing it.
- **`spmd_bootstrap` initialised MPI on every import.** `mpi4py` is a declared dependency, so the
  `try: from mpi4py import MPI` always succeeded, and importing it calls `MPI_Init` as a side
  effect — in single-process runs, in pytest workers, and in CI where no MPI runtime is
  configured. It now detects an MPI launcher from the environment and returns before importing
  anything when there is none. Under a launcher with `mpi4py` missing it raises instead of letting
  every rank silently run the whole problem.
- **`derivative_coeffs`** now computed through cumulative sum algorithm because the alternative scan is poor on gpu.

Known limitation, not fixed: `spmd_bootstrap` binds the coordinator to `127.0.0.1`, which is fine
for a single node but cannot work under a multi-node `mpirun`.

### CODE REVIEW

**Comment by mikaem** [comment]
@augustfe Sorry about the size of the PR. Started optimizing and found it difficult to stop. In the end the examples/OrrSommerfel.py problem runs at 1100 steps/s on my macbook. Compared with 70 steps/s for the identical Shenfun solver (not shared memory). Also runs at 700 steps/s on a single A100 GPU (Saga). Not really tested much in scale yet (multiple devices), but works and speeds up.

File: examples/ChannelFlow2D.py
Comment (augustfe) [inline examples/ChannelFlow2D.py:492]: There may be some value in implementing a "dumb" version of this, with a lower barrier-to-entry. You would have to already be quite specialized in order to follow the remainder of the script

File: examples/RayleighBenard.py [hunk 1]
```diff
@@ -302,11 +317,16 @@ def scalar_initial(self) -> tuple[Array, ...]:
...
             noise = jax.random.normal(jax.random.PRNGKey(self.seed), Vh.shape)
-        else:
+            T_hat = Vh.forward(self.amplitude * noise * (1 - yy**2))
+        elif self.mode == 1:
             noise = jnp.cos(2 * jnp.pi * self.mode * xx / self.Lx) * jnp.ones_like(yy)
-        T_hat = Vh.forward(self.amplitude * noise * (1 - yy**2))
+            T_hat = Vh.forward(self.amplitude * noise * (1 - yy**2))
+        else:
+            noise = jax.random.uniform(jax.random.PRNGKey(self.seed), self.VT.shape)
+            T_hat = self.VT.forward(self.amplitude * noise * (1 - yy**2))
```
Comment (augustfe) [inline examples/RayleighBenard.py:328]: Is `self.VT.forward` different from `Vh.forward`?

File: src/jaxfun/galerkin/Chebyshev.py [hunk 2]
```diff
@@ -41,6 +52,28 @@ def _dense_derivative_matrix(
...
+    """Return W4 with the Chebyshev forward transform's scaling folded in."""
+    c = (-1.0) ** np.arange(N) / N
+    c[0] *= 0.5
+    return _dct_twiddle(N) * c
+
+
+@cache
+def _chebyshev_scalar_product_scale(N: int, domain_factor: float) -> np.ndarray:
+    """Return W4 with the Chebyshev scalar product's scaling folded in."""
+    return _dct_twiddle(N) * (np.pi * (-1.0) ** np.arange(N) / (2 * N * domain_factor))
```
Comment (augustfe) [inline src/jaxfun/galerkin/Chebyshev.py:74]: Haven't gotten to `_dct_twiddle` yet, but what is the value of caching vs jit (maybe with static args?) in this case

File: src/jaxfun/integrators/etdrk4.py
Comment (augustfe) [inline src/jaxfun/integrators/etdrk4.py:1]: May still need to update the contour integral here

File: src/jaxfun/la/matrix.py [hunk 3]
```diff
@@ -487,13 +488,16 @@ def _pin(
...
+        # Preserve the value's own type: floats stay floats (unaffected repr),
+        # complex values are kept as complex rather than truncated.
+        norm_constraints: list[tuple[int, complex]] = []
         for idx, val in constraints:
             if idx < -n or idx >= n:
                 raise IndexError(
                     f"Constraint index {idx} is out of range for matrix of size {n}"
                 )
-            norm_constraints.append((idx % n, float(val)))
+            val = complex(val) if isinstance(val, complex) else float(val)
```
Comment (augustfe) [inline src/jaxfun/la/matrix.py:499]: What's the point of this line? If complex, cast to complex?

File: src/jaxfun/la/tpmatrix.py [hunk 4]
```diff
@@ -28,6 +35,18 @@
...
 
 
+def _sharding():
+    """Return `jaxfun.sharding`, imported on use rather than at module scope.
+
+    `jaxfun.sharding` imports `jaxfun.typing`, which imports `jaxfun.la`, which
+    is this package -- so importing it from here at module scope closes a cycle.
+    Every caller below is on a multi-device path, long after imports settle.
+    """
+    from jaxfun import sharding
```
Comment (augustfe) [inline src/jaxfun/la/tpmatrix.py:45]: Another great chance to wait for `lazy from jaxfun import sharding`

File: src/jaxfun/utils/common.py [hunk 5]
```diff
@@ -243,37 +246,264 @@ def reverse_dict[K, V](d: dict[K, V]) -> dict[V, K]:
...
+# The twiddles are built on the host and memoized, for two separate reasons. XLA
+# does not constant-fold the `arange`/`exp` that makes them, even though N is
+# static, and leaving them inside the jit costs 1.7x. And they must be numpy
+# rather than jnp: these are reached lazily from inside a trace, and on a
+# distributed run from inside a `shard_map`, where `jnp.asarray` hands back a
+# tracer that the cache would then store and every later call from outside that
+# trace would leak. Host arrays carry no trace or mesh affiliation and XLA embeds
+# them as literals, which is also why `cache_static` goes via the host. Being
+# literals they need no communication when distributed: each device gets its own
+# copy compiled in, and the module picks up no collectives from them.
```
Comment (augustfe) [inline src/jaxfun/utils/common.py:391]: aha, I see

**Review by augustfe — APPROVED** [review]
Looks very good! Time to implement the OpenAI Navier--Stokes counter example next?

File: examples/ChannelFlow2D.py
Comment (mikaem) [reply examples/ChannelFlow2D.py:492]: Yes, but unfortunately dumb version does not exist, because we here mix a 2D transient pde with a 1D transient pde. The algorithm needs to be explained, sure, but for now this particular algorithm needs low-level stuff. Cannot simply formulate equations and press integrate:-(

File: examples/RayleighBenard.py (hunk 1 above)
Comment (mikaem) [reply examples/RayleighBenard.py:328]: Yes. VT is inhomogeneous, Vh is homogeneous. The VT path creates an initial condition that is zero+noise everywhere, except T=1 at the lower wall. The other two initial conditions start from a linear profile in T (plus noise) going from 1 at the lower wall to 0 at the top.

File: src/jaxfun/galerkin/Chebyshev.py (hunk 2 above)
Comment (mikaem) [reply src/jaxfun/galerkin/Chebyshev.py:74]: Haven't measured anything yet, but these are all called from within jit, and it seamed like a good idea to cache the produced arrays that only depend on the length N. These are called over and over in forward and backward transforms over each Fourier wavenumber. Always with same N. But maybe it makes no difference?

File: src/jaxfun/utils/common.py (hunk 5 above)
Comment (mikaem) [reply src/jaxfun/utils/common.py:391]: Aha!
